### PR TITLE
Initial look at transient material inverse optimization

### PR DESCRIPTION
**Pull request by zachmprince**
Refs #22215

### CODE REVIEW

File: modules/optimization/examples/materialTransient/optimize_grad.i
Comment (lynnmunday) [inline modules/optimization/examples/materialTransient/optimize_grad.i:1]: add forward_out_data_0011.csv

File: modules/optimization/examples/materialTransient/forward.i [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  []
+[]
+
+[Reporters]
+  [diffc_rep]
+    type = ConstantReporter
+    real_vector_names = 'D_x_coord D_y_coord D_vals'
+    real_vector_values = '0.25 0.75 0.25 0.75;
+                          0.25 0.25 0.75 0.75;
+                          1  0.1   0.1   0.01' # Reference solution
```
Comment (mguddati) [inline modules/optimization/examples/materialTransient/forward.i:67]: Change this to 1  0.2   0.2   0.05' # Reference solution

File: modules/optimization/examples/materialTransient/optimize_grad.i [hunk 2]
```diff
@@ -0,0 +1,88 @@
+[Optimization]
+[]
+
+[OptimizationReporter]
+  type = OptimizationReporter
+  parameter_names = 'D'
+  num_values = '4'
+  initial_condition = '1 1 1 1'
```
Comment (mguddati) [inline modules/optimization/examples/materialTransient/optimize_grad.i:8]: Change lines 8-10 to:
  initial_condition = '0.01 0.01 0.01 0.01'
  upper_bounds = '1e2'
  lower_bounds = '1e-3'

**Review by mguddati — CHANGES_REQUESTED** [review]
Zach - the problem you have is somewhat stiff (with two orders of magnitude difference in the parameter range), which is somewhat challenging. I made it a bit mellow (1.0 and 0.05 are the extreme true values of parameters). In addition, the starting point optimization makes a difference and it appears best to start with a lower bound then from above. So, I changed the starting point to small parameter values. The gradient optimization now converges.

File: modules/optimization/examples/materialTransient/forward.i (hunk 1 above)
Comment (zachmprince) [reply modules/optimization/examples/materialTransient/forward.i]: Great advice! Updated

File: modules/optimization/examples/materialTransient/optimize_grad.i (hunk 2 above)
Comment (zachmprince) [reply modules/optimization/examples/materialTransient/optimize_grad.i]: Updated!

File: modules/optimization/examples/materialTransient/optimize_grad.i
Comment (zachmprince) [reply modules/optimization/examples/materialTransient/optimize_grad.i:1]: Running `forward.i Outputs/execute_on=FINAL` will produce this file

File: framework/src/userobjects/SolutionUserObject.C [hunk 3]
```diff
@@ -262,6 +262,9 @@ SolutionUserObject::readExodusII()
 
   // Build nodal/elemental variable lists, limit to variables listed in 'system_variables', if
   // provided
+  _nodal_variables.clear();
+  _elemental_variables.clear();
+  _scalar_variables.clear();
```
Comment (lindsayad) [inline framework/src/userobjects/SolutionUserObject.C:267]: Why is this change necessary? A comment might be good

Comment (zachmprince) [reply framework/src/userobjects/SolutionUserObject.C:269]: I have  derived class that is calling this function multiple times. So I need to clear these before doing the following `push_back`s. I've added a comment.

Comment (lindsayad) [reply framework/src/userobjects/SolutionUserObject.C:269]: thank you!

**Review by lynnmunday — DISMISSED** [review]

File: modules/optimization/examples/materialTransient/gold/forward_out_data_0011.csv [hunk 4]
```diff
@@ -0,0 +1,41 @@
+measurement_time,measurement_values,measurement_xcoord,measurement_ycoord,measurement_zcoord,misfit_values,simulation_values
+0.1,0,0.25,0.25,0,0.098572025012904,0.098572025012904
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/examples/materialTransient/gold/forward_out_data_0011.csv:2]: This is just the result of running `forward.i` alone so I can get the simulated measurement data. The actual optimization results are in the exodus gold files.

File: modules/optimization/src/vectorpostprocessors/OptimizationFunctionInnerProductHelper.C [hunk 5]
```diff
@@ -102,7 +102,7 @@ OptimizationFunctionInnerProductHelper::getVector(std::vector<Real> & result)
       it.second.resize(nvar);
     result.assign(nvar, 0.0);
 
-    // Integrate in time using quadrature rule
+    // Integrate in time
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/src/vectorpostprocessors/OptimizationFunctionInnerProductHelper.C:105]: The main reason it's a map is so that if this object is called twice on the same time step it overwrites the data instead of appending a vector. So no, it still needs to be a map.

File: modules/optimization/src/userobjects/AdjointSolutionUserObject.C [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    SolutionUserObject::initialSetup();
+
+    // Make sure to communicate what solution was actually loaded
+    _actual_interpolation_time = 0.0;
+    _interpolation_time = -_dt;
+  }
+
+  // Update time interpolation for ExodusII solution
+  const Real actual_time = _reverse_time_end - _t + _dt;
+  if (_actual_interpolation_time != actual_time)
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/src/userobjects/AdjointSolutionUserObject.C:75]: It makes things easier to allow interpolation since there might be floating point discrepancies. This also allows for some customization on how much data is kept from the forward problem. Maybe we can reduce the number of time steps outputted from the forward problem to save memory and do an approximate adjoint.

[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/userobjects/AdjointSolutionUserObject.C:75]: will you click resolved for my conversations?  I can't do it.

**Comment by lynnmunday** [comment]
How do we merge your commits?  I can't do it

File: modules/optimization/src/userobjects/AdjointSolutionUserObject.C [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  if (stats.st_mtime > _file_mod_time)
+  {
+    _file_mod_time = stats.st_mtime;
+
+    // Tell the parent class that we do need to re-initialize
+    _initialized = false;
+    // EquationSystems doesn't like the destructor that's called when there is a
+    // unique_ptr::operator=
+    _es.reset();
+    _es2.reset();
```
Comment (dschwen) [inline modules/optimization/src/userobjects/AdjointSolutionUserObject.C:63]: Why would that be a different dtor than when you're calling reset?

**Review by dschwen — APPROVED** [review]
Looks good to me.

File: modules/optimization/src/userobjects/AdjointSolutionUserObject.C (hunk 7 above)
Comment (zachmprince) [reply modules/optimization/src/userobjects/AdjointSolutionUserObject.C:63]: I'm really not sure. I thought the behavior strange too.
